TAS5805M + BridgeMode + Player-Settings

DerPicknicker opened the pull request:
Hello @CarlosDerSeher , 

as discussed here is the new PR. We can now switch the Inversion of the LogicLevels inside the Menuconfig: 
<img width="719" alt="image" src="https://github.com/CarlosDerSeher/snapclient/assets/64746593/0431eeba-d169-40cb-9a1d-2f9ae9c13e1e">

The Bridge Mode is only shown if TAS5805M is selected: 
<img width="374" alt="image" src="https://github.com/CarlosDerSeher/snapclient/assets/64746593/37839911-d89c-4b1b-906d-b8edaa17227e">


I have a small bug which I can't find. Please be so kind and check it. 
```
Driver param setup : 0
Driver installed   : 0
W (638) TAS5805M: Power down pin: 33
W (1041) TAS5805M: Setting to HI Z
E (1141) AUDIO_HAL: codec init failed!
I (1141) HAL: codec_hal done
E (1141) AUDIO_BOARD: board.c:82 (audio_board_codec_init): Got NULL Pointer
I (1145) AUDIO_BOARD: board-handle done
I (1150) SC: Audio board_init done
E (1154) AUDIO_HAL: audio_hal handle is null
E (1159) AUDIO_HAL: audio_hal handle is null
I (1164) SC: init player
I (1167) PLAYER: player_setup_i2s: dma_buf_len is 576, dma_buf_count is 4
W (1174) i2s_common: i2s controller 0 has been occupied by i2s_driver
E (1181) i2s_common: i2s_new_channel(812): no available channel found
ESP_ERROR_CHECK failed: esp_err_t 0x105 (ESP_ERR_NOT_FOUND) at 0x400fe978
0x400fe978: player_setup_i2s at /project/snapclient/components/lightsnapcast/player.c:213

file: "./components/lightsnapcast/player.c" line 213
func: player_setup_i2s
expression: i2s_new_channel(&tx_chan_cfg, &tx_chan, NULL)

abort() was called at PC 0x4008a20b on core 0
0x4008a20b: _esp_error_check_failed at /opt/esp/idf/components/esp_system/esp_err.c:50
```

Thank you in Advance. I think the APLL & and LogicLevels are great enhancement also for later debugging purposes

CarlosDerSeher commented:
Please format the logs as code.

> LogicLevels inside the Menuconfig:

Shouldn't these be with i2s pin settings?

DerPicknicker commented:
> Please format the logs as code.
Fixed.

> Shouldn't these be with i2s pin settings?

I am flexible - I can move it if you want.

CarlosDerSeher commented:
> AUDIO_HAL: codec init failed!

Why is it failing?

I wouldszggest placing the i2s signal inversion with pin config. And please name it invert instead of flip

CarlosDerSeher commented on components/custom_board/tas5805m/tas5805m.c:
```diff
@@ -0,0 +1,277 @@
+  gpio_set_level(TAS5805M_GPIO_PDN, 0);
+  vTaskDelay(200 / portTICK_RATE_MS);
+  gpio_set_level(TAS5805M_GPIO_PDN, 1);
+  vTaskDelay(200 / portTICK_RATE_MS);
```
Is this really needed?

CarlosDerSeher replied:
@DerPicknicker do these delays really need to be 100s of ms? Shouldn't these be around a couple us?

CarlosDerSeher commented on components/custom_board/tas5805m/tas5805m.c:
```diff
@@ -0,0 +1,277 @@
+  ESP_LOGW(TAG, "Power down pin: %d", TAS5805M_GPIO_PDN);
+  gpio_config(&io_conf);
+  gpio_set_level(TAS5805M_GPIO_PDN, 0);
+  vTaskDelay(200 / portTICK_RATE_MS);
```
Is this really needed seems quite long

CarlosDerSeher replied:
@DerPicknicker do these delays really need to be 100s of ms? Shouldn't these be around a couple us?

CarlosDerSeher commented on components/custom_board/tas5805m/tas5805m.c:
```diff
@@ -0,0 +1,277 @@
+
+  ESP_LOGW(TAG, "Setting to HI Z");
+  ret = tas5805m_write_byte(TAS5805M_DEVICE_CTRL_2_REGISTER, 0x02);
+  vTaskDelay(100 / portTICK_RATE_MS);
```
Needed?

CarlosDerSeher commented on components/custom_board/tas5805m/tas5805m.c:
```diff
@@ -0,0 +1,277 @@
+    return ret;
+#endif
+
+  vTaskDelay(100 / portTICK_RATE_MS);
```
Needed?

DerPicknicker replied:
@CarlosDerSeher .. I will reduce them to 50micro seconds. Hope it works then anyways.

CarlosDerSeher commented on components/custom_board/tas5805m/tas5805m.c:
```diff
@@ -0,0 +1,277 @@
+  res = i2c_driver_install(i2c_master_port, i2c_cfg.mode,
+                           I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE,
+                           0);
+  printf("Driver installed   : %d\n", res);
```
Please use ESP LOG always

DerPicknicker replied:
Is done in the next commit.

CarlosDerSeher replied:
Did you change this already?

DerPicknicker replied:
I think here I messed up while debugging. I will change it to ESP_LOGI again. I was happy to validate that the idf5 i2s driver works with the inversion so I don't keep my eyes on that ;-)

CarlosDerSeher commented on components/lightsnapcast/Kconfig.projbuild:
```diff
@@ -0,0 +1,18 @@
+config FLIP_BCLK_LEVEL
+            bool "Invert Logic Level BCLK"
+            default false
+config DISABLE_APLL_ADJUST
```
Please add some description what this is doing, e.g. disable sync implementation!

DerPicknicker replied:
Is done in the next commits. I will push them soon.

CarlosDerSeher replied:
@DerPicknicker I don't like that you can disable sync implementation. Also I changed your merge request to sample stuffing branch which will include a choice for apll  s sample stuffing. Please remove this setting

DerPicknicker replied:
@CarlosDerSeher  Just remove the setting inside the Menuconfig or also the line in the code where I use it? I guess the second option ;-)

DerPicknicker commented:
I will reduce the delays to 50% all of them are needed for the dac. Under IDF4 the code runs fine. I copy the files one by one. Compile and resolve the missing reference and config values. 

How can I check why the Init fails? There is an error check which says there is a problem inside the player.c file. Do you have any suggestions what can cause such a behaviour?

CarlosDerSeher commented:
Are these delays really ms? I'd assume us

CarlosDerSeher commented:
https://github.com/CarlosDerSeher/snapclient/blob/master/components%2Faudio_hal%2Faudio_hal.c#L70  I'd really check your init function and where exactly the error is generated. Seems to me i2c isn't working as expected maybe. Check `ret`

DerPicknicker commented:
Yes. Might be true. I can't measure it. 

I will change the position and test it a bit more. Would you be so kind and check the issue around to player.c? I can't see things which can cause this behaviour.

DerPicknicker commented:
For me it's weird it worked under IDF 4 it's crazy how espressif change their apis and break compatibility... I will compare the files one by one if I made a mistake.

DerPicknicker commented:
Hello @CarlosDerSeher .. 

I did some debugging. The Issue is related to this piece of Code: 
```
ret = i2c_master_cmd_begin(I2C_TAS5805M_MASTER_NUM, cmd, 1000 / portTICK_RATE_MS);
  if (ret != ESP_OK) {
    ESP_LOGE(TAG, "Error at master_cmd_begin: %s", esp_err_to_name(ret));
  }
```
I checked the Documentation but I don't understand what I should change to make it work. I refer to this Documentation: https://demo-dijiudu.readthedocs.io/en/latest/api-reference/peripherals/i2c.html#_CPPv216i2c_param_config10i2c_port_tPK12i2c_config_t

I changed the Postion of the LogicLevels and Renamed them.

DerPicknicker commented:
The clock rate works fine under IDF4. I will try to check before it. I added some error check macros and this work. I will check it if the pins are correct. 

I used the default values and I used 1:1 the same code. I will compare sdkconfigs again. 

Do you have any other ideas?

CarlosDerSeher commented:
Currently not really. You could try writing and reading some register and verify communication...

CarlosDerSeher commented:
Next step would probably be to look at the i2c signals. But yeah strange it is working on IDF 4

DerPicknicker commented:
@CarlosDerSeher .. I found out that the default values are somehow ignored (of the menuconfig) ... 

So the I2C pins aren't correct. I will correct them and test it again. I will override them manually in the menuconfig and give you an update. Thinking about the pins was too trivial for me because I set the default values correct. Strange.....

CarlosDerSeher commented:
Hmm maybe something was changed in kconfig from v4 to v5?

DerPicknicker commented:
I used your sdkconfig and selected my dac. But somehow it won't be saved. I will test it tomorrow. If it works we can merge it hopefully. 

BTW: Do you know if we can allow SPDIF output on the esp32 like here: https://github.com/amedes/esp_a2dp_sink_spdif

CarlosDerSeher commented:
> BTW: Do you know if we can allow SPDIF output on the esp32 like here: https://github.com/amedes/esp_a2dp_sink_spdif

Probably yes but not at the same time with analog audio I guess?!

DerPicknicker commented:
Hi @CarlosDerSeher .. 

I really can't understand what causes this issue. Changing the i2C Pins and enabling PSRAM got me one step ahead. 

But I really can't understand why the Code with the old SDK-Config works fine. Should I send you the old and new config for checking? 

Here is the log: 
```
I (29) boot: ESP-IDF v5.1.1 2nd stage bootloader
I (29) boot: compile time Jan 30 2024 15:36:03
I (29) boot: Multicore bootloader
I (33) boot: chip revision: v3.1
I (37) qio_mode: Enabling default flash chip QIO
I (43) boot.esp32: SPI Speed      : 80MHz
I (47) boot.esp32: SPI Mode       : QIO
I (52) boot.esp32: SPI Flash Size : 4MB
I (56) boot: Enabling RNG early entropy source...
I (62) boot: Partition Table:
I (65) boot: ## Label            Usage          Type ST Offset   Length
I (73) boot:  0 nvs              WiFi data        01 02 00009000 00004000
I (80) boot:  1 otadata          OTA data         01 00 0000d000 00002000
I (88) boot:  2 phy_init         RF data          01 01 0000f000 00001000
I (95) boot:  3 factory          factory app      00 00 00010000 00120000
I (102) boot:  4 ota_0            OTA app          00 10 00130000 00120000
I (110) boot:  5 ota_1            OTA app          00 11 00250000 00120000
I (118) boot:  6 storage          Unknown data     01 82 00370000 00090000
I (125) boot: End of partition table
I (129) boot: Defaulting to factory image
I (134) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=31498h (201880) map
I (198) esp_image: segment 1: paddr=000414c0 vaddr=3ffb0000 size=03740h ( 14144) load
I (203) esp_image: segment 2: paddr=00044c08 vaddr=40080000 size=0b410h ( 46096) load
I (219) esp_image: segment 3: paddr=00050020 vaddr=400d0020 size=c7650h (816720) map
I (444) esp_image: segment 4: paddr=00117678 vaddr=4008b410 size=09a34h ( 39476) load
I (467) boot: Loaded app from partition at offset 0x10000
I (468) boot: Disabling RNG early entropy source...
I (479) cpu_start: Multicore app
I (479) quad_psram: This chip is ESP32-D0WD
I (479) esp_psram: Found 8MB PSRAM device
I (481) esp_psram: Speed: 40MHz
I (485) esp_psram: PSRAM initialized, cache is in low/high (2-core) mode.
W (493) esp_psram: Virtual address not enough for PSRAM, map as much as we can. 4MB is mapped
I (502) cpu_start: Pro cpu up.
I (505) cpu_start: Starting app cpu, entry point is 0x400814e8
0x400814e8: call_start_cpu1 at /opt/esp/idf/components/esp_system/port/cpu_start.c:154

I (498) cpu_start: App cpu up.
I (1370) esp_psram: SPI SRAM memory test OK
I (1377) cpu_start: Pro cpu start user code
I (1377) cpu_start: cpu freq: 240000000 Hz
I (1377) cpu_start: Application information:
I (1381) cpu_start: Project name:     snapclient
I (1386) cpu_start: App version:      4c9b1ed-dirty
I (1391) cpu_start: Compile time:     Jan 30 2024 15:51:10
I (1398) cpu_start: ELF file SHA256:  3896591a9183ac37...
I (1404) cpu_start: ESP-IDF:          v5.1.1
I (1409) cpu_start: Min chip rev:     v0.0
I (1413) cpu_start: Max chip rev:     v3.99 
I (1418) cpu_start: Chip rev:         v3.1
I (1423) heap_init: Initializing. RAM available for dynamic allocation:
I (1431) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (1437) heap_init: At 3FFB9EC0 len 00026140 (152 KiB): DRAM
I (1443) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (1449) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (1456) heap_init: At 40094E44 len 0000B1BC (44 KiB): IRAM
I (1462) esp_psram: Adding pool of 4096K of PSRAM memory to heap allocator
I (1470) spi_flash: detected chip: generic
I (1474) spi_flash: flash io: qio
I (1479) app_start: Starting scheduler on CPU0
I (1484) app_start: Starting scheduler on CPU1
I (1484) main_task: Started on CPU0
I (1494) esp_psram: Reserving pool of 32K of internal memory for DMA/internal allocations
I (1502) main_task: Calling app_main()
I (1535) SC: Start codec chip
I (1535) HAL: INIT
W (1536) TAS5805M: Power down pin: 33
W (1736) TAS5805M: Setting to HI Z
W (1836) TAS5805M: Setting to PLAY
I (1836) HAL: codec_hal done
I (1836) AUDIO_BOARD: board-handle done
I (1837) SC: Audio board_init done
I (1841) AUDIO_HAL: Codec mode is 3, Ctrl:1
I (1846) SC: init player
I (1849) PLAYER: player_setup_i2s: dma_buf_len is 576, dma_buf_count is 4
W (1856) i2s_common: i2s controller 0 has been occupied by i2s_driver
E (1863) i2s_common: i2s_new_channel(812): no available channel found
ESP_ERROR_CHECK failed: esp_err_t 0x105 (ESP_ERR_NOT_FOUND) at 0x400f79e0
0x400f79e0: player_setup_i2s at /project/snapclient/components/lightsnapcast/player.c:213

file: "./components/lightsnapcast/player.c" line 213
func: player_setup_i2s
expression: i2s_new_channel(&tx_chan_cfg, &tx_chan, NULL)

abort() was called at PC 0x4008d3a7 on core 0
0x4008d3a7: _esp_error_check_failed at /opt/esp/idf/components/esp_system/esp_err.c:50



Backtrace: 0x40081dcd:0x3ffbb550 0x4008d3b1:0x3ffbb570 0x40091721:0x3ffbb590 0x4008d3a7:0x3ffbb600 0x400f79e0:0x3ffbb630 0x400f7d2b:0x3ffbb6e0 0x400da7c3:0x3ffbb730 0x40193272:0x3ffbb7e0
0x40081dcd: panic_abort at /opt/esp/idf/components/esp_system/panic.c:452

0x4008d3b1: esp_system_abort at /opt/esp/idf/components/esp_system/port/esp_system_chip.c:84

0x40091721: abort at /opt/esp/idf/components/newlib/abort.c:38

0x4008d3a7: _esp_error_check_failed at /opt/esp/idf/components/esp_system/esp_err.c:50

0x400f79e0: player_setup_i2s at /project/snapclient/components/lightsnapcast/player.c:213

0x400f7d2b: init_player at /project/snapclient/components/lightsnapcast/player.c:360

0x400da7c3: app_main at /project/snapclient/main/main.c:2847

0x40193272: main_task at /opt/esp/idf/components/freertos/app_startup.c:208 (discriminator 13)


```

CarlosDerSeher commented:
Ok, maybe the problem is related to this https://github.com/CarlosDerSeher/snapclient/blob/bd17be1b7fc40ca888a9e58d5a0b395e8128b928/main/main.c#L2782
 When using a custom board I enable i2s and occupy the channel, but it is never freed, at least not with the handle aquired in main. Please remove this temporary for testing if your DAC doesn't need MCLK during initialization

DerPicknicker commented:
@CarlosDerSeher .. 

I tested your requested Idea and it works almost fine. The Log has some issues which aren't there in IDF4 so this should be also considered: 

```
W (15854) PLAYER: RESYNCING HARD 2: age -718329us, latency 620504193396us, free 4186127, largest block 3997696, 29, rssi: -55
E (15855) gptimer: gptimer_stop(356): timer is not running
E (15861) gptimer: gptimer_stop(356): timer is not running
E (15867) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
I (16652) PLAYER: initial sync age: 14us, chunk duration: 24000us
I (28796) SC: Buffer length:  1500
I (28796) SC: Latency:        0
I (28796) SC: Mute:           0
I (28797) SC: Setting volume: 20
I (29196) SC: Buffer length:  1500
I (29196) SC: Latency:        0
I (29196) SC: Mute:           0
I (29196) SC: Setting volume: 22
I (30571) SC: Buffer length:  1500
I (30571) SC: Latency:        0
I (30571) SC: Mute:           0
I (30572) SC: Setting volume: 20
I (31135) SC: Buffer length:  1500
I (31135) SC: Latency:        0
I (31135) SC: Mute:           0
I (31135) SC: Setting volume: 18
I (34811) SC: fLaC sampleformat: 48000:16:2
W (36234) PLAYER: RESYNCING HARD 2: age 5327us, latency 620504193212us, free 4340591, largest block 4128768, 0, rssi: -53
E (36235) gptimer: gptimer_stop(356): timer is not running
E (38241) PLAYER: Couldn't get PCM chunk, recv: messages waiting 0, diff2Server: 620504s, 193.212ms
E (38241) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
E (40248) PLAYER: Couldn't get PCM chunk, recv: messages waiting 0, diff2Server: 620504s, 193.212ms
E (40248) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
I (41095) SC: fLaC sampleformat: 48000:16:2
E (41939) gptimer: gptimer_stop(356): timer is not running
E (41939) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
I (43305) PLAYER: initial sync age: 10us, chunk duration: 24000us
I (58064) SC: fLaC sampleformat: 48000:16:2
I (58218) SC: fLaC sampleformat: 48000:16:2
W (60654) PLAYER: RESYNCING HARD 2: age -876676us, latency 620504192987us, free 4218667, largest block 4063232, 22, rssi: -58
E (60656) gptimer: gptimer_stop(356): timer is not running
E (60662) gptimer: gptimer_stop(356): timer is not running
E (60668) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet
```

CarlosDerSeher commented:
> E (60662) gptimer: gptimer_stop(356): timer is not running
E (60668) i2s_common: i2s_channel_disable(1030): the channel has not been enabled yet

I know these are there. Not very beautiful and also misleading but they didn't cause any issues for me. Still I will look into it at some point

DerPicknicker commented:
Okay, good to know. I will verify the code if it works also if the invert logic works. I will try an test mp3 for testing left/right channels if this works as expected I will give you here an update. I will also remove a lot of unnecessary code and provide an SDK_Config for that board.

DerPicknicker commented:
@CarlosDerSeher should I adapt the code with an fixed #if which enables your fixes if my DAC is selected or you will change the code later? 

What should I do before merging into idf5? I will try to test it tomorrow if my code works as expected.

CarlosDerSeher commented:
Maybe just  change it to `#if 0` for now. As far as I know the i2s setup in main is only necessary for my custom built pcb's dac. I'll change it later. I'd like to call the setup function from main, so the handles  used are the same.

DerPicknicker commented:
I will do it.

DerPicknicker commented:
Hello @CarlosDerSeher ... 

Sorry for the Delay. I tested now my Code and from my side everything works as expected. 

I tested the Bridge-Mode, Word-Select-Inversion and it works fine. The Channels are inverted by default because the Board plays without inversion the wrong channel. 

I added also a SDK-Config which works fine on my side.

DerPicknicker commented:
Hi, 

sorry for the delay. I removed some of the old code and changed the delay from 100 / 200 to 10. This should be enough.

DerPicknicker commented:
Hi @CarlosDerSeher .. Did you found the Time for checking this PR?

CarlosDerSeher commented:
No there are some merge conflicts. Maybe you could try and resolve these locally at your end? I'd like to merge to sync_with_sample_stuffing

CarlosDerSeher commented:
Please remove the apll adjust thing and we are ready to merge.

Regarding merging sample insert with master, I really want to do do it soon but you said it, I couldn't find the time yet. Weekends are short and the evenings too :)

DerPicknicker commented:
Hi, 

I removed the APLL-Stuff and some other unused code. I tried the Actions in my fork but I have no luck until it's merged in the master branch. Because I can't run the Action not inside a different Branch as the master. 

Regarding the Web-Server it would be good if we can made the changes after the build is done this would allow to build it once and made the changes after the GitHub action is finished. So nobody has to setup the build env again if he wants to change some settings. But we should discuss it in the discussion thread ;-)

CarlosDerSeher commented:
FInally! Maybe I can find a few hours next week to get all this into master.
